File: WorldFeed/Server/src/Services/Profile/WorldFeed.Profile.API/Infrastructure/Inject/TwitterClient.cs
// ReSharper disable once CheckNamespace
namespace WorldFeed.Profile.Infrastructure.Inject
{
    using System;

    using WorldFeed.Common.Events;
    using WorldFeed.Common.Models.Authentication;
    using WorldFeed.Common.Public.Client.Clients;
    using WorldFeed.Common.Public.Events;
    using WorldFeed.Common.Settings;
    using WorldFeed.Profile.Application.Requesters;
    using WorldFeed.Profile.Client;

    public class TwitterClientParameters
    {
        public TwitterClientParameters()
        {
            Settings = new TweetinviSettings();
        }

        public IRateLimitCache RateLimitCache { get; set; }
        public IProfileContainer Container { get; set; }
        public ITweetinviSettings Settings { get; set; }
        public event EventHandler<ProfileContainerEventArgs> BeforeRegistrationCompletes;

        public void RaiseBeforeRegistrationCompletes(ProfileContainerEventArgs args)
        {
            args.ProfileContainer.Raise(BeforeRegistrationCompletes, args);
        }
    }

    public class TwitterClient : ITwitterClient
    {
        private IReadOnlyTwitterCredentials _credentials;
        private readonly IProfileContainer _tweetinviContainer;
        private readonly ITwitterClientEvents _twitterClientEvents;

        /// <summary>
        /// IMPORTANT NOTE: The setter is for convenience. It is strongly recommended to create a new TwitterClient instead.
        /// As using this setter could result in unexpected concurrency between the time of set and the execution of previous
        /// non awaited async operations.
        /// </summary>
        public IReadOnlyTwitterCredentials Credentials
        {
            get => _credentials;
            set => _credentials = new ReadOnlyTwitterCredentials(value);
        }

        public ITweetinviSettings Config { get; }

        public TwitterClient(IReadOnlyConsumerCredentials credentials) : this(new ReadOnlyTwitterCredentials(credentials), new TwitterClientParameters())
        {
        }

        public TwitterClient(IReadOnlyTwitterCredentials credentials) : this(credentials, new TwitterClientParameters())
        {
        }

        public TwitterClient(string consumerKey, string consumerSecret) : this(new ReadOnlyTwitterCredentials(consumerKey, consumerSecret), new TwitterClientParameters())
        {
        }

        public TwitterClient(string consumerKey, string consumerSecret, string bearerToken) : this(new ReadOnlyTwitterCredentials(consumerKey, consumerSecret, bearerToken), new TwitterClientParameters())
        {
        }

        public TwitterClient(string consumerKey, string consumerSecret, string accessToken, string accessSecret) : this(new ReadOnlyTwitterCredentials(consumerKey, consumerSecret, accessToken, accessSecret), new TwitterClientParameters())
        {
        }

        public TwitterClient(IReadOnlyTwitterCredentials credentials, TwitterClientParameters parameters)
        {
            Credentials = credentials;
            Config = parameters?.Settings ?? new TweetinviSettings();

            if (parameters?.Container == null)
            {
                if (!Infrastructure.ProfileContainer.Container.IsInitialized)
                {
                    Infrastructure.ProfileContainer.Container.Initialize();
                }
            }
            else
            {
                if (!parameters.Container.IsInitialized)
                {
                    throw new InvalidOperationException("Cannot create a client with a non initialized container!");
                }
            }

            _tweetinviContainer = new ProfileContainer(parameters?.Container ?? Infrastructure.ProfileContainer.Container);
            _tweetinviContainer.RegisterInstance(typeof(IProfileContainer), _tweetinviContainer);

            if (parameters?.RateLimitCache != null)
            {
                _tweetinviContainer.RegisterInstance(typeof(IRateLimitCache), parameters.RateLimitCache);
            }

            _tweetinviContainer.RegisterInstance(typeof(TwitterClient), this);
            _tweetinviContainer.RegisterInstance(typeof(ITwitterClient), this);

            void BeforeRegistrationDelegate(object sender, TweetinviContainerEventArgs args)
            {
                parameters?.RaiseBeforeRegistrationCompletes(args);
            }

            _tweetinviContainer.BeforeRegistrationCompletes += BeforeRegistrationDelegate;
            _tweetinviContainer.Initialize();
            _tweetinviContainer.BeforeRegistrationCompletes -= BeforeRegistrationDelegate;

            var requestExecutor = _tweetinviContainer.Resolve<IRawExecutors>();
            Raw = requestExecutor;

            var parametersValidator = _tweetinviContainer.Resolve<IParametersValidator>();
            ParametersValidator = parametersValidator;

            Execute = _tweetinviContainer.Resolve<IExecuteClient>();
            RateLimits = _tweetinviContainer.Resolve<IRateLimitsClient>();
            Tweets = _tweetinviContainer.Resolve<ITweetsClient>();
            Users = _tweetinviContainer.Resolve<IUsersClient>();

            _tweetinviContainer.AssociatedClient = this;

            _twitterClientEvents = _tweetinviContainer.Resolve<ITwitterClientEvents>();
            Factories = _tweetinviContainer.Resolve<ITwitterClientFactories>();
            Json = _tweetinviContainer.Resolve<IJsonClient>();

            var rateLimitCacheManager = _tweetinviContainer.Resolve<IRateLimitCacheManager>();
            rateLimitCacheManager.RateLimitsClient = RateLimits;
        }

        /// <inheritdoc/>
        public IExecuteClient Execute { get; }

        /// <inheritdoc/>
        public IRateLimitsClient RateLimits { get; }

        /// <inheritdoc/>
        public IUsersClient Users { get; }

        /// <inheritdoc/>
        public IExternalClientEvents Events => _twitterClientEvents;
        /// <inheritdoc/>
        public ITwitterClientFactories Factories { get; }
        /// <inheritdoc/>
        public IJsonClient Json { get; }

        /// <inheritdoc/>
        public IParametersValidator ParametersValidator { get; }
        /// <inheritdoc/>
        public IRawExecutors Raw { get; }

        public ITwitterExecutionContext CreateTwitterExecutionContext()
        {
            return new TwitterExecutionContext
            {
                RequestFactory = CreateRequest,
                Container = _tweetinviContainer,
                Events = _twitterClientEvents
            };
        }

        public ITwitterRequest CreateRequest()
        {
            var request = new TwitterRequest
            {
                ExecutionContext = CreateTwitterExecutionContext(),
                Query =
                {
                    // we are cloning here to ensure that the context will never be modified regardless of concurrency
                    TwitterCredentials = new TwitterCredentials(Credentials),
                }
            };

            request.Query.Initialize(Config);
            request.ExecutionContext.Initialize(Config);

            return request;
        }
    }
}

File: Sharebook/Client/src/app/sharebook/TwitterClient.ts
import {Inject, Injectable, Injector, Optional, ReflectiveInjector} from "@angular/core";

import {ISharebookSettings, SharebookSettings} from "../core/Public/Settings/SharebookSettings";
import {IRateLimitCache} from "../core/Core/RateLimit/IRateLimitCache";
import {ITwitterClient, ITwitterClientToken} from "../core/Public/ITwitterClient";
import {
  IReadOnlyTwitterCredentials,
  ReadOnlyTwitterCredentials
} from '../core/Core/Models/Authentication/ReadOnlyTwitterCredentials';
import {ITwitterClientFactories} from "../core/Public/Client/Tools/ITwitterClientFactories";
import {IParametersValidator, ParametersValidator} from "../core/Core/Client/Validators/ParametersValidator";
import {ITwitterExecutionContext, TwitterExecutionContext} from "../core/Core/Client/TwitterExecutionContext";
import {IRawExecutors, IRawExecutorsToken} from "../core/Public/Client/IRawExecutors";
import {IAuthClient} from "../core/Public/Client/Clients/IAuthClient";
import {IAccountSettingsClient} from "../core/Public/Client/Clients/IAccountSettingsClient";
import {IExecuteClient} from "../core/Public/Client/Clients/IExecuteClient";
import {IHelpClient} from "../core/Public/Client/Clients/IHelpClient";
import {IListsClient} from "../core/Public/Client/Clients/IListsClient";
import {IMessagesClient} from "../core/Public/Client/Clients/IMessagesClient";
import {IRateLimitsClient} from "../core/Public/Client/Clients/IRateLimitsClient";
import {ISearchClient} from "../core/Public/Client/Clients/ISearchClient";
import {IStreamsClient} from "../core/Public/Client/Clients/IStreamsClient";
import {ITimelinesClient} from "../core/Public/Client/Clients/ITimelinesClient";
import {ITrendsClient} from "../core/Public/Client/Clients/ITrendsClient";
import {ITweetsClient} from "../core/Public/Client/Clients/ITweetsClient";
import {IUploadClient} from "../core/Public/Client/Clients/IUploadClient";
import {IUsersClient} from "../core/Public/Client/Clients/IUsersClient";
import {IAccountActivityClient} from "../core/Public/Client/Clients/IAccountActivityClient";
import {ITwitterRequest} from "../core/Public/Models/Interfaces/ITwitterRequest";
import {TwitterRequest} from "../core/Public/TwitterRequest";
import {TwitterQuery} from "../core/Public/TwitterQuery";
import {TwitterCredentials} from "../core/Public/Models/Authentication/TwitterCredentials";
import {IExternalClientEvents, ITwitterClientEvents} from "../core/Core/Events/TweetinviGlobalEvents";
import {IJsonClient} from "../core/Public/Client/Clients/IJsonClient";
import {UploadClient} from "./Client/Clients/UploadClient";
import {RawExecutors} from "./Client/RawExecutors";
import {IAccountActivityRequester} from "../core/Public/Client/Requesters/IAccountActivityRequester";
import {AccountActivityRequester} from "./Client/Requesters/AccountActivityRequester";
import {IAccountSettingsRequester} from "../core/Public/Client/Requesters/IAccountSettingsRequester";
import {IHelpRequester} from "../core/Public/Client/Requesters/IHelpRequester";
import {ISearchRequester} from "../core/Public/Client/Requesters/ISearchRequester";
import {ITwitterListsRequester} from "../core/Public/Client/Requesters/ITwitterListsRequester";
import {AuthRequester} from "./Client/Requesters/AuthRequester";
import {AccountSettingsRequester} from "./Client/Requesters/AccountSettingsRequester";
import {HelpRequester} from "./Client/Requesters/HelpRequester";
import {SearchRequester} from "./Client/Requesters/SearchRequester";
import {TwitterListsRequester} from "./Client/Requesters/TwitterListsRequester";
import {TimelinesRequester} from "./Client/Requesters/TimelinesRequester";
import {TrendsRequester} from "./Client/Requesters/TrendsRequester";
import {TweetsRequester} from "./Client/Requesters/TweetsRequester";
import {UploadRequester} from "./Client/Requesters/UploadRequester";
import {UsersRequester} from "./Client/Requesters/UsersRequester";
import {AuthClient} from "./Client/Clients/AuthClient";
import {AccountSettingsClient} from "./Client/Clients/AccountSettingsClient";
import {ExecuteClient} from "./Client/Clients/ExecuteClient";
import {HelpClient} from "./Client/Clients/HelpClient";
import {ListsClient} from "./Client/Clients/ListsClient";
import {MessagesClient} from "./Client/Clients/MessagesClient";
import {RateLimitsClient} from "./Client/Clients/RateLimitsClient";
import {SearchClient} from "./Client/Clients/SearchClient";
import {TimelinesClient} from "./Client/Clients/TimelinesClient";
import {TrendsClient} from "./Client/Clients/TrendsClient";
import {TweetsClient} from "./Client/Clients/TweetsClient";
import {UsersClient} from "./Client/Clients/UsersClient";
import {AccountActivityClient} from "./Client/Clients/AccountActivityClient";
import {AppInjector} from "./Injectinvi/app-injector";
import {IRateLimitCacheManager} from "../core/Core/RateLimit/IRateLimitCacheManager";
import {RateLimitCacheManager} from "../Tweetinvi.Credentials/RateLimit/RateLimitCacheManager";
import {App} from "../core/Core/Models/Properties/App";

@Injectable({
  providedIn: 'root',
})
export class TwitterClientParameters {
  constructor() {
    this.settings = new SharebookSettings();
  }

  public rateLimitCache: IRateLimitCache;

  public settings: ISharebookSettings;

  // public event BeforeRegistrationCompletes: EventHandler<TweetinviContainerEventArgs>;

  // public RaiseBeforeRegistrationCompletes(args: TweetinviContainerEventArg): void {
  //   args.TweetinviContainer.Raise(BeforeRegistrationCompletes, args);
  // }
}

@Injectable({
  providedIn: 'root',
})
export class TwitterClient implements ITwitterClient {
  private _credentials: IReadOnlyTwitterCredentials;
  // private readonly _tweetinviContainer: ITweetinviContainer;
  private readonly _twitterClientEvents: ITwitterClientEvents;

  constructor(parameters?: TwitterClientParameters) {
    // this.credentials = credentials;
    this.config = parameters?.settings ?? new SharebookSettings();

    // if (parameters?.rateLimitCache != null) {
    //   _tweetinviContainer.RegisterInstance(typeof(IRateLimitCache), parameters.RateLimitCache);
    // }

    // Injector.create({
    //   providers: [{provide: TwitterClient, useValue: this}]
    // });
    // Injector.create({
    //   providers: [{provide: ITwitterClientToken, useValue: this}]
    // });
    // _tweetinviContainer.RegisterInstance(typeof(TwitterClient), this);
    // _tweetinviContainer.RegisterInstance(typeof(ITwitterClient), this);

    // Injector.create({providers: [{provide: TwitterClient, useValue: this}]});

    // debugger
    // let requestExecutor = this.injector.get(IRawExecutorsToken); // _tweetinviContainer.Resolve<IRawExecutors>();
    // this.raw = requestExecutor;

    // let parametersValidator = this._tweetinviContainer.Resolve<IParametersValidator>();
    // this.parametersValidator = parametersValidator;

    // this.auth = this._tweetinviContainer.Resolve<IAuthClient>();
    // this.accountSettings = this._tweetinviContainer.Resolve<IAccountSettingsClient>();
    // this.execute = this._tweetinviContainer.Resolve<IExecuteClient>();
    // this.help = this._tweetinviContainer.Resolve<IHelpClient>();
    // this.lists = this._tweetinviContainer.Resolve<IListsClient>();
    // this.messages = this._tweetinviContainer.Resolve<IMessagesClient>();
    // this.rateLimits = AppInjector.get(RateLimitsClient); // this._tweetinviContainer.Resolve<IRateLimitsClient>();
    // this.search = this._tweetinviContainer.Resolve<ISearchClient>();
    // this.streams = this._tweetinviContainer.Resolve<IStreamsClient>();
    // this.timelines = this._tweetinviContainer.Resolve<ITimelinesClient>();
    // this.trends = this._tweetinviContainer.Resolve<ITrendsClient>();
    // this.tweets = this._tweetinviContainer.Resolve<ITweetsClient>();
    // this.upload = new UploadClient(this); // this._tweetinviContainer.Resolve<IUploadClient>();
    // this.users = this._tweetinviContainer.Resolve<IUsersClient>();
    // this.accountActivity = this._tweetinviContainer.Resolve<IAccountActivityClient>();

    // this._tweetinviContainer.AssociatedClient = this;

    // this._twitterClientEvents = this._tweetinviContainer.Resolve<ITwitterClientEvents>();
    // this.factories = this._tweetinviContainer.Resolve<ITwitterClientFactories>();
    // this.json = this._tweetinviContainer.Resolve<IJsonClient>();

    // debugger
    // let rateLimitCacheManager = AppInjector.get(RateLimitCacheManager);
    // rateLimitCacheManager.rateLimitsClient = this.rateLimits;
  }

  // IMPORTANT NOTE: The setter is for convenience. It is strongly recommended to create a new TwitterClient instead.
  // As using this setter could result in unexpected concurrency between the time of set and the execution of previous
  // non awaited async operations.
  get credentials(): IReadOnlyTwitterCredentials {
    return this._credentials;
  }

  set credentials(value: IReadOnlyTwitterCredentials) {
    this._credentials = new ReadOnlyTwitterCredentials(value, null);
  }


  public config: ISharebookSettings;

  public auth: IAuthClient;
  public accountSettings: IAccountSettingsClient;
  public execute: IExecuteClient;
  public help: IHelpClient;
  public lists: IListsClient;
  public messages: IMessagesClient;
  public rateLimits: IRateLimitsClient;
  public search: ISearchClient;
  public streams: IStreamsClient;
  public timelines: ITimelinesClient;
  public trends: ITrendsClient;
  public tweets: ITweetsClient;
  public upload: IUploadClient;
  public users: IUsersClient;
  public accountActivity: IAccountActivityClient;

  get events(): IExternalClientEvents {
    return this._twitterClientEvents;
  }

  public factories: ITwitterClientFactories;
  public json: IJsonClient;

  public parametersValidator: IParametersValidator;
  public raw: IRawExecutors;

  public createTwitterExecutionContext(): ITwitterExecutionContext {
    let result = new TwitterExecutionContext();
    result.requestFactory = this.createRequest;
    // result.container = this._tweetinviContainer;
    result.events = this._twitterClientEvents;

    return result;
  }

  public createRequest(): ITwitterRequest {
    let twitterQuery = new TwitterQuery();
    // we are cloning here to ensure that the context will never be modified regardless of concurrency
    // @ts-ignore
    twitterQuery.twitterCredentials = new TwitterCredentials(this.credentials);

    let request = new TwitterRequest();
    request.executionContext = this.createTwitterExecutionContext();
    request.query = twitterQuery;

    request.query.initialize(this.config);
    request.executionContext.initialize(this.config);

    return request;
  }
}
